File: codecheck-case/cpp/BKCT_BackwardNull/BKCT_BackwardNull_01.cpp
#include <iostream>

using namespace std;


void testNullCheck() {
    char *str = nullptr;
    //bad
	// NOCC: BackwardNull(检查工具误报)
    cout << str[0] << endl;
    if (str != nullptr) {
        cout << "String is not null" << endl;
    }
}

void anotherTest() {
    int *ptr = nullptr;
    //bad
    cout << *ptr << endl;
    if (ptr != nullptr) {
        cout << "Pointer is not null" << endl;
    }
}

void invalidAccess() {
    int *arr = nullptr;
    //bad
    cout << arr[0] << endl;
    if (arr != nullptr) {
        cout << "Array pointer is not null" << endl;
    }
}

void badAccess() {
    class MyObject {
    public:
        void methodCall() {
            cout << "Method called" << endl;
        }
    };

    MyObject *myObj = nullptr;
    //bad
    myObj->methodCall();
    if (myObj != nullptr) {
        cout << "MyObject pointer is not null" << endl;
    }
}


void testNullCheckGood() {
    char *str = nullptr;
    if (str != nullptr) {
        //good
        cout << str[0] << endl;
    }
}

void anotherTestGood() {
    int *ptr = nullptr;
    if (ptr != nullptr) {
        //good
        cout << *ptr << endl;
    }
}

void validAccess() {
    int *arr = nullptr;
    if (arr != nullptr) {
        //good
        cout << arr[0] << endl;
    }
}

void goodAccess() {
    class MyObject {
    public:
        void methodCall() {
            cout << "Method called" << endl;
        }
    };

    MyObject *myObj = nullptr;
    if (myObj != nullptr) {
        //good
        myObj->methodCall();
    }
}

int main() {
    // run some test functions if needed
    return 0;
}

File: codecheck-case/lua/BKCT_BackwardNull/BKCT_BackwardNull_01.lua

function testNullCheck()
    local str = nil
    -- bad
    print(#str)
    if str ~= nil then
        print("String is not nil")
    end
end

function anotherTest()
    local ptr = nil
    -- bad
    print(ptr.someProperty)
    if ptr ~= nil then
        print("Pointer is not nil")
    end
end

function invalidAccess()
    local arr = nil
    -- bad
    print(arr[1])
    if arr ~= nil then
        print("Array is not nil")
    end
end

function badAccess()
    local myObj = nil
    -- bad
    myObj:methodCall()
    if myObj ~= nil then
        print("MyObject is not nil")
    end
end


function testNullCheckGood()
    local str = nil
    if str ~= nil then
        -- good
        print(#str)
    end
end

function anotherTestGood()
    local ptr = nil
    if ptr ~= nil then
        -- good
        print(ptr.someProperty)
    end
end

function validAccess()
    local arr = nil
    if arr ~= nil then
        -- good
        print(arr[1])
    end
end

function goodAccess()
    local myObj = nil
    if myObj ~= nil then
        -- good
        myObj:methodCall()
    end
end

-- Method definition for testing
function MyObject:methodCall()
    print("Method called")
end